File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque

n, m = map(int, input().split())

graph = [[] * (0) for i in range(m)]


for i in range(m):
    graph[i] = list(map(int, input().split()))

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

q = deque()

# 토마토가 있는 곳 찾기
for i in range(m):
    for j in range(n):
        if graph[i][j] == 1:
            q.append((i, j))


def bfs():

    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx >= 0 and ny >= 0 and nx < m and ny < n:
                if graph[nx][ny] == 0:
                    q.append((nx, ny))
                    graph[nx][ny] = graph[x][y] + 1

bfs()

day = 0

for row in graph:
    for tomato in row:
        if tomato == 0:
            print(-1)
            exit(0)
    day = max(day, max(row))

print(day - 1)

File: 백준/Gold/7576. 토마토/토마토.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static int N, M, max;
    static int[][] graph;
//    static boolean[][] visited;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static Queue<Tomato> q;


    static class Tomato {
        int x, y, day;

        public Tomato(int x, int y, int day) {
            this.x = x;
            this.y = y;
            this.day = day;
        }

    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        graph = new int[M][N];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < N; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        max = 0;
        q = new LinkedList<>();

        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (graph[i][j] == 1) {
                    q.offer(new Tomato(i, j, 0));
                }
            }
        }

        bfs();

        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                if (graph[i][j] == 0) {
                    max = -1;
                }
            }
        }

        System.out.println(max);
    }

    private static void bfs() {

        while(!q.isEmpty()) {
            Tomato tomato = q.poll();
            int x = tomato.x;
            int y = tomato.y;
            max = Math.max(max, tomato.day);

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx >= 0 && ny >= 0 && nx < M && ny < N) {
                    if (graph[nx][ny] == 0) {
                        q.offer(new Tomato(nx, ny, tomato.day + 1));
                        graph[nx][ny] = 1;
                    }
                }
            }
        }

    }

}
